src/test/java/com/google/devtools/build/lib/skylark/cpp/SkylarkToolchainTest.java

```diff
@@ -0,0 +1,217 @@
+      "    return",
+      "  fail('Lists do not match {} <> {}'.format(list1, list2))",
+      "def assertTrue(arg):",
+      "  if not bool(arg):",
+      "    fail('Value is %s. Expected convertible to True.' % arg)");
+  }
+
+  @Test
+  public void testEscaping() throws Exception {
+    scratch.file(
+        "test/skylark/extension.bzl",
+        "load('//tools/cpp:lib_cc_configure.bzl', 'escape_string', 'unescape_string')",
+        "load('//test/skylark:empty.bzl', 'empty')",
+        "def test(arg, **kwargs):",
+        "  escaped = escape_string(arg)",
+        "  result = unescape_string(escaped)",
+        "  if result != arg:",
+        "    fail('Could not reverse escaping {} -escape> {} -unescape> {}'.format(arg, escaped, result))",
+        "  return empty(**kwargs)");
+
+    scratch.file(
+        "test/skylark/BUILD",
+        "load('//test/skylark:extension.bzl', 'test')",
+        "test('\\my%simple;%%test!%%%for?%%escaping:%works/', name = 'test_target')");
+
+    getTarget("//test/skylark:test_target");
+  }
+
+  @Test
+  public void testError() throws Exception {
+    scratch.file(
+        "test/skylark/extension.bzl",
+        "load('//tools/cpp:lib_cc_configure.bzl', 'get_escaping_error')",
+        "load('//test/skylark:assert.bzl', 'assertIsNone', 'assertTrue')",
+        "load('//test/skylark:empty.bzl', 'empty')",
+        "def test(**kwargs):",
+        "  error = get_escaping_error('foo')",
+        "  assertIsNone(error)",
+        "  error = get_escaping_error('foo;', additionals = ';')",
+        "  assertTrue(error)",
+        "  error = get_escaping_error('foo%a')",
+        "  assertTrue(error)",
+        "  error = get_escaping_error('foo%')",
+        "  assertTrue(error)",
+        "  error = get_escaping_error('foo%;bar', additionals = ';')",
+        "  assertIsNone(error)",
+        "  empty(**kwargs)");
+
+    scratch.file(
+        "test/skylark/BUILD",
+        "load('//test/skylark:extension.bzl', 'test')",
+        "test(name = 'test_target')");
+
+    getTarget("//test/skylark:test_target");
+  }
+
+  class FailHandler implements EventHandler {
+     public String failString = null;
+      @Override
+      public void handle(Event event) {
+        if (EventKind.ERRORS.contains(event.getKind())) {
+          failString = event.toString();
+        }
+      }
+  }
+
+  @Test
+  public void testErrorFail() throws Exception {
+
+    scratch.file(
+      "test/skylark/extension.bzl",
+      "load('//tools/cpp:lib_cc_configure.bzl', 'get_escaping_error')",
+      "load('//test/skylark:empty.bzl', 'empty')",
+      "def test(**kwargs):",
+      "  get_escaping_error('foo', additionals = '%')",
+      "  empty(**kwargs)");
+
+    scratch.file(
+        "test/skylark/BUILD",
+        "load('//test/skylark/extension.bzl', 'test')",
+        "test(name = 'test_target')");
+    
+    ConfiguredTarget target = getConfiguredTarget("//test/skylark:test_target");
+    assertThat(view.hasErrors(target)).isTrue();
```

Comment:
assertContainsEvent("...")

you might need to say `reporter.removeHandler(failFastHandler);`

Reply:
Sorry, I do not follow.

Reply:
`BuildViewTestCase` has an `assertContainsEvent` method that checks that certain errors are reported. Please use it here.

Reply:
Tried to but it does not work (`eventCollector` stays empty). Not really sure why. I fear `Reporter` may filter the failure. Since the Event handling seems to be hackish as is, I do not want to make it event more so.
So I would very much like to keep the `hasErrors`.